src/ray/object_manager/object_manager.cc

```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
 }
 
 ray::Status ObjectManager::Cancel(const ObjectID &object_id) {
-  ray::Status status = object_directory_->UnsubscribeObjectLocations(object_id);
+  ray::Status status = object_directory_->UnsubscribeObjectLocations(
+      object_directory_pull_callback_id_, object_id);
   return status;
 }
 
-ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids,
-                                uint64_t timeout_ms, int num_ready_objects,
+ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids, int64_t wait_ms,
+                                uint64_t num_required_objects, bool wait_local,
                                 const WaitCallback &callback) {
-  // TODO: Implement wait.
+  UniqueID wait_id = UniqueID::from_random();
+
+  if (wait_local) {
+    return ray::Status::NotImplemented("Wait for local objects is not yet implemented.");
+  }
+
+  RAY_CHECK(wait_ms >= 0);
+  RAY_CHECK(num_required_objects != 0);
+  RAY_CHECK(num_required_objects <= object_ids.size());
+  if (object_ids.size() == 0) {
+    callback(std::vector<ObjectID>(), std::vector<ObjectID>());
+  }
+
+  // Initialize fields.
+  active_wait_requests_.emplace(wait_id, WaitState(*main_service_, wait_ms, callback));
+  auto &wait_state = active_wait_requests_.find(wait_id)->second;
+  wait_state.object_id_order = object_ids;
+  wait_state.wait_ms = wait_ms;
+  wait_state.num_required_objects = num_required_objects;
+  for (auto &oid : object_ids) {
+    if (local_objects_.count(oid) > 0) {
+      wait_state.found.insert(oid);
+    } else {
+      wait_state.remaining.insert(oid);
+    }
+  }
+
+  if (wait_state.remaining.empty()) {
+    WaitComplete(wait_id);
+  } else {
+    for (auto &oid : wait_state.remaining) {
+      // Lookup remaining objects.
+      wait_state.requested_objects.insert(oid);
+      RAY_CHECK_OK(object_directory_->LookupLocations(
+          oid, [this, wait_id](const std::vector<ClientID> &client_ids,
+                               const ObjectID &object_id) {
+            auto &wait_state = active_wait_requests_.find(wait_id)->second;
+            if (!client_ids.empty()) {
+              wait_state.remaining.erase(object_id);
+              wait_state.found.insert(object_id);
+            }
```

Comment:
What would it mean if `client_ids.empty() == true`? That would mean that the object has been evicted? How do we want to handle that case?

Reply:
It may mean that the object has been evicted. It may also mean that no entries exist for the object. In both cases, after the round of lookups, if we need to wait for more objects, `SubscribeObjectLocations` will be invoked on objects for which `client_ids.empty() == true` (in `AllWaitLookupsComplete`). The handler to `SubscribeObjectLocations` will be invoked when `!client_ids.empty()`.